src/call/client.rs

```diff
@@ -0,0 +1,346 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use futures::{Async, AsyncSink, Future, Poll, Sink, StartSend, Stream};
+use grpc_sys;
+use protobuf::{self, Message, MessageStatic};
+
+use async::{BatchMessage, CqFuture};
+use call::Call;
+use error::Error;
+use super::{SinkBase, StreamingBase};
+
+/// Update the flag bit in res.
+#[inline]
+fn change_flag(res: &mut u32, flag: u32, set: bool) {
+    if set {
+        *res |= flag;
+    } else {
+        *res &= !flag;
+    }
+}
+
+#[derive(Default)]
+pub struct CallOption {
+    timeout: Option<Duration>,
+    write_flags: u32,
+    call_flags: u32,
+}
+
+impl CallOption {
+    /// Signal that the call is idempotent
+    pub fn with_idempotent(mut self, is_idempotent: bool) -> CallOption {
+        change_flag(&mut self.call_flags,
+                    grpc_sys::GRPC_INITIAL_METADATA_IDEMPOTENT_REQUEST,
+                    is_idempotent);
+        self
+    }
+
+    /// Signal that the call should not return UNAVAILABLE before it has started
+    pub fn with_wait_for_ready(mut self, wait_for_ready: bool) -> CallOption {
+        change_flag(&mut self.call_flags,
+                    grpc_sys::GRPC_INITIAL_METADATA_WAIT_FOR_READY,
+                    wait_for_ready);
+        self
+    }
+
+    /// Signal that the call is cacheable. GRPC is free to use GET verb
+    pub fn with_cacheable(mut self, cacheable: bool) -> CallOption {
+        change_flag(&mut self.call_flags,
+                    grpc_sys::GRPC_INITIAL_METADATA_CACHEABLE_REQUEST,
+                    cacheable);
+        self
+    }
+
+    /// Hint that the write may be buffered and need not go out on the wire immediately.
+    pub fn with_buffer_hint(mut self, need_buffered: bool) -> CallOption {
+        change_flag(&mut self.write_flags,
+                    grpc_sys::GRPC_WRITE_BUFFER_HINT,
+                    need_buffered);
+        self
+    }
+
+    /// Force compression to be disabled.
+    pub fn with_force_no_compress(mut self, no_compress: bool) -> CallOption {
+        change_flag(&mut self.write_flags,
+                    grpc_sys::GRPC_WRITE_NO_COMPRESS,
+                    no_compress);
+        self
+    }
+
+    /// Set a timeout.
+    pub fn with_timeout(mut self, timeout: Duration) -> CallOption {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    /// Get the timeout.
+    pub fn timeout(&self) -> Option<Duration> {
+        self.timeout
+    }
+}
+
+/// A handler to handle uanry async call.
+///
+/// The future is resolved once response is received.
+pub struct UnaryCallHandler<T> {
+    call: Call,
+    resp_f: CqFuture<BatchMessage>,
+    _resp: PhantomData<T>,
```

Comment:
I don't think it's true. Name it `_resp` here to indicate that T is the type of response.